design_notes/v2.md

```diff
@@ -0,0 +1,98 @@
+# VirtualiZarr V2 architecture discussion
+
+## Goals
+
+- Extensible and easily understandable entry-point system for new backends
+- Modular code base that uses Obspec for the object storage interface abstraction, Zarr as the core data model, and Xarray for the concatenation and merging of Zarr stores
+
+## Related issues
+
+- [Fragility of url auto-parsing logic](https://github.com/zarr-developers/VirtualiZarr/issues/561)
+- [Use obstore / obspec for globbing in `open_virtual_mfdataset`](https://github.com/zarr-developers/VirtualiZarr/issues/569)
+- [Function definition for open_virtual_dataset in next major release](https://github.com/zarr-developers/VirtualiZarr/issues/553)
+- [Support passing configuration options to default_object_store](https://github.com/zarr-developers/VirtualiZarr/issues/559)
+- [Redefine Virtual Readers as func(filepath) -> ManifestStore](https://github.com/zarr-developers/VirtualiZarr/issues/498)
+- [Make xarray an optional dependency?](https://github.com/zarr-developers/VirtualiZarr/issues/521)
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: An object defining the paths, chunk byte offsets, lengths of one virtual array.
```

Comment:
Should `ChunkManifest` be changed to `ManifestChunk` (It sounds terrible I know) to match the `ManifestArray`, `ManifestGroup` and `ManifestStore` naming convention?

Reply:
If we wanted to make them consistent I think we would instead have just `Manifest`, or change the others to be `ChunkManifestArray` etc. But I don't really think this matters very much.

Reply:
`ChunkManifest` is the only object that really directly contains a manifest (of chunks, in this case). If we were to rename these, I would chose `ChunkManifest`, `VirtualArray`, `VirtualGroup`, and `VirtualStore`.

Reply:
> `ChunkManifest` is the only object that really directly contains a manifest (of chunks, in this case). If we were to rename these, I would chose `ChunkManifest`, `VirtualArray`, `VirtualGroup`, and `VirtualStore`.

FWIW I like these names

Reply:
> Is the only object that really directly contains a manifest (of chunks, in this case). 

That's not true - all of `ManifestArray`, `ManifestGroup`, `ManifestStore` contain manifests, they are just wrapped up in lower-level objects. Containing manifests is their defining feature!

If anything I regret the "virtual" nomenclature, not the "manifest" nomenclature.

Reply:
Literally a `ManifestArray` is defined by only two attributes (`._metadata` and `._manifest`), and `zarr.Array` also has `._metadata`.

Reply:
Perhaps "contain" wasn't the correct word to use. My point is that ChunkManifest **is** a Manifest, so it makes sense to include "Manifest" in the name. In contrast, the others **have** a manifest.

For example, the most important feature of `ManifestStore` is that it redirects an incoming key and byte range to a different key and byte range. IMO "VirtualStore" or "IndirectionStore" (same message, more boring) is much more aligned to why it exists.

This doesn't matter a lot so I'm fine with just leaving it all, but I think the "virtual" nomenclature is more clear for objects that don't have storing references as their primary purpose.

Reply:
I can see the argument for having everything in the library be "virtual-something" though...